src/hotspot/share/opto/predicates.cpp

```diff
@@ -147,3 +148,331 @@ Node* PredicateBlock::skip_regular_predicates(Node* regular_predicate_proj, Deop
   }
   return entry;
 }
+
+TemplateAssertionPredicateBool::TemplateAssertionPredicateBool(Node* source_bool) : _source_bool(source_bool->as_Bool()) {
+#ifdef ASSERT
+  // During IGVN, we could have multiple outputs of the _source_bool, for example, when the backedge of the loop of
+  // this Template Assertion Predicate is about to die and the CastII on the last value bool already folded to a
+  // constant (i.e. no OpaqueLoop* nodes anymore). Then IGVN could already have commoned up the bool with the bool of
+  // one of the Hoisted Check Predicates. Just check that the Template Assertion Predicate is one of the outputs.
+  bool has_template_output = false;
+  for (DUIterator_Fast imax, i = source_bool->fast_outs(imax); i < imax; i++) {
+    Node* out = source_bool->fast_out(i);
+    if (out->Opcode() == Op_Opaque4) {
+      has_template_output = true;
+      break;
+    }
+  }
+  assert(has_template_output, "must find Template Assertion Predicate as output");
+#endif // ASSERT
+}
+
+// Stack used when performing DFS on a Template Assertion Predicate Bool. The DFS traversal visits non-CFG inputs of a
+// node in increasing node index order (i.e. first visiting the input node at index 1). Each time a new node is visited,
+// it is inserted on top of the stack. Each node n in the stack maintains a node input index i, denoted as [n, i]:
+//
+// i = 0, [n, 0]:
+//     n is currently being visited for the first time in the DFS traversal and was newly added to the stack
+//
+// i > 0, [n, i]:
+//     Let node s be the next node being visited after node n in the DFS traversal. The following holds:
+//         n->in(i) = s
+class DFSNodeStack : public StackObj {
+  Node_Stack _stack;
+  static const uint _no_inputs_visited_yet = 0;
+
+ public:
+  explicit DFSNodeStack(BoolNode* template_bool)
+      : _stack(2) {
+        _stack.push(template_bool, _no_inputs_visited_yet);
+  }
+
+  // Push the next unvisited input of the current node on top of the stack and return true. The visiting order of node
+  // inputs is in increasing node input index order. If there are no unvisited inputs left, do nothing and return false.
+  bool push_next_unvisited_input() {
+    Node* node_on_top = top();
+    increment_top_node_input_index();
+    const uint next_unvisited_input = _stack.index();
+    for (uint index = next_unvisited_input; index < node_on_top->req(); index++) {
+      Node* input = node_on_top->in(index);
+      if (TemplateAssertionPredicateBool::could_be_part(input)) {
+        // We only care about nodes that could possibly be part of a Template Assertion Predicate Bool.
+        _stack.set_index(index);
+        _stack.push(input, _no_inputs_visited_yet);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  Node* top() const {
+    return _stack.node();
+  }
+
+  uint node_index_to_previously_visited_parent() const {
```

Comment:
This is the "parent" of the current node, right?
Parent is also a bit confusing, I think. Maybe use "input" instead? Because at the beginning you say the DFS traverses the inputs.
Why do you say "previously visited"?

Reply:
Ah. I think it should simply be `top_input_index`. You could then also rename `top` -> `top_node`. And `increment_top_node_input_index` -> `increment_top_index`, if you even think it is worth keeping.